ksql-engine/src/main/java/io/confluent/ksql/planner/PlanSourceExtractorVisitor.java

```diff
@@ -0,0 +1,80 @@
+/**
+ * Copyright 2017 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+package io.confluent.ksql.planner;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import io.confluent.ksql.planner.plan.AggregateNode;
+import io.confluent.ksql.planner.plan.FilterNode;
+import io.confluent.ksql.planner.plan.JoinNode;
+import io.confluent.ksql.planner.plan.OutputNode;
+import io.confluent.ksql.planner.plan.PlanNode;
+import io.confluent.ksql.planner.plan.PlanVisitor;
+import io.confluent.ksql.planner.plan.ProjectNode;
+import io.confluent.ksql.planner.plan.StructuredDataSourceNode;
+
+public class PlanSourceExtractorVisitor<C, R> extends PlanVisitor<C, R> {
+
+  final Set<String> sourceNames;
+
+  public PlanSourceExtractorVisitor() {
+    sourceNames = new HashSet<>();
+  }
+
+  public R process(PlanNode node, C context) {
+    return node.accept(this, context);
+  }
+
+  protected R visitPlan(PlanNode node, C context) {
+    return null;
+  }
+
+  protected R visitFilter(FilterNode node, C context) {
```

Comment:
I'm a bit weary of us hardcoding the assumption that a node has a single source. I understand that we would only ever have one source for anything but a join, and its unlikely we will ever have more, but I think it would be better to enforce that somewhere upstream while not assuming that here and iterating over the list. That way if we do decide we want to support merges we don't have all these little changes that can potentially get missed.

Reply:
We do indeed have a set for source for a node. Every node should implement `public abstract List<PlanNode> getSources();`.

Reply:
Shouldn't this iterate over all the sources then?

Reply:
I'm a +1 on this. Why bother having a `List<>` if we only ever access the first item in the list?

Reply:
Depending on the node type you will have one or more sources. For instance for `Filter` node you will always have single source but for `Join` node you can have two or more sources.